wagtail/core/blocks/base.py

```diff
@@ -107,33 +113,62 @@ def set_name(self, name):
     def media(self):
         return forms.Media()
 
-    def html_declarations(self):
-        """
-        Return an HTML fragment to be rendered on the form page once per block definition -
-        as opposed to once per occurrence of the block. For example, the block definition
-            ListBlock(label="Shopping list", CharBlock(label="Product"))
-        needs to output a <script type="text/template"></script> block containing the HTML for
-        a 'product' text input, to that these can be dynamically added to the list. This
-        template block must only occur once in the page, even if there are multiple 'shopping list'
-        blocks on the page.
+    def get_layout(self):
+        return self.SIMPLE
 
-        Any element IDs used in this HTML fragment must begin with definition_prefix.
-        (More precisely, they must either be definition_prefix itself, or begin with definition_prefix
-        followed by a '-' character)
-        """
-        return ''
+    def prepare_value(self, value, errors=None):
+        return value
 
-    def js_initializer(self):
-        """
-        Returns a Javascript expression string, or None if this block does not require any
-        Javascript behaviour. This expression evaluates to an initializer function, a function that
-        takes the ID prefix and applies JS behaviour to the block instance with that value and prefix.
+    def prepare_for_react(self, parent_block, value,
+                          type_name=None, errors=None):
+        if type_name is None:
+            type_name = self.name
+        from .stream_block import StreamValue
+        if isinstance(value, StreamValue.StreamChild):
```

Comment:
This is raising major alarm bells. It's absolutely essential to the design of StreamField that the `Block` class should define a contract for each block type to follow, _independently of any other block's implementation_. That way, we can verify the code by checking that each block implementation is following that contract, rather than having to consider every possible interaction between different parent/child block types.

No block other than `StreamBlock` should ever have to know about the `StreamChild` class. Additionally, there should never be any need for a block to vary its behaviour based on the identity of its parent, as we're doing here (`if parent_block is None`).

I suspect there is a similar logic error to https://github.com/wagtail/wagtail/pull/4481/files#r190932257 here: a `StreamBlock` value is a sequence of _wrapped_ child values (i.e. wrapped in a `StreamChild` object in the Python representation, and wrapped in a `{'type': ..., 'value': ..., 'id': ...}` dict in the JSON representation). `StreamBlock`'s `prepare_value` / `prepare_for_react` methods should be responsible for unwrapping each child value before passing it to the child block's `prepare_for_react` - that way, `Block.prepare_for_react` will consistently receive the unwrapped value and not need a special case for `StreamChild` - and `StreamBlock` will also be responsible for re-wrapping the returned value in a `BlockData` object.

Once all of the `StreamBlock`-specific logic here has been moved into the `StreamBlock` class, I believe it will reduce to:

    def prepare_for_react(self, value, errors=None):
        return self.prepare_value(value, errors=errors)

...which brings us back to the question of why we need both `prepare_value` and `prepare_for_react`, I guess :-)

Reply:
That’s a good discussion to have. The main reason for this design is compatibility, as you will see. Maybe it can be improved so that there is no `StreamChild` reference, but I’m not sure how.

In theory, we should  not need to differenciate between `prepare_for_react` and `prepare_value`, as you pointed out.

I made that difference is because currently, the database representation of StreamFields is not really consistent.
Here is a comparison between database representation and react-streamfield representation, so you can understand the purpose of this:

StreamBlock of FieldBlocks in database:
```JSON
[{"id": "98a82", "type": "a", "value": 1}, {"id": "87b1", "type": "b", "value": 2}]
```

StreamBlock in react-streamfield:
```JSON
[{"id": "98a82", "type": "a", "value": 1}, {"id": "87b1", "type": "b", "value": 2}]
```

ListBlock in database:
```JSON
[1, 2]
```

ListBlock in react-streamfield:
```JSON
[{"id": "98a82", "type": "a", "value": 1}, {"id": "87b1", "type": "b", "value": 2}]
```

StructBlock in database:
```JSON
{"id": "98a82", "type": "person", "value": {"first_name": "Bertrand", "last_name": "Bordage"}}
```

StructBlock in react-streamfield:
```JSON
{"id": "b198e", "type": "person", "value": [
    {"id": "98a82", "type": "first_name", "value": "Bertrand"},
    {"id": "87b1", "type": "last_name", "value": "Bordage"}
]}
```

You can notice two main things here:
- In react-streamfield, all blocks are represented as an object, with an ID and a type. they can also have their own `hasError` and `html` keys, but that’s not really important here.
- In react-streamfield, ListBlock is exactly the same thing as a StreamBlock. Not only in terms of representation here, but also internally.

So as you can see, the poor FieldBlocks are not always considered as blocks in the database representation. The way they are treated depends on their parent. When a FieldBlock is in a StreamBlock, it’s considered as a normal block. When it’s in a ListBlock, it’s considered as a value. When it’s a StructBlock, it’s considered as a key/value pair.

I appreciate that the database representation is minimal. It doesn’t add more data than is required, which is good. But it’s truely a pain to parse because of this.

Also, there was no way I could just send the database representation to react-streamfield. We need more information than just a value in a good JavaScript StreamField. We need an `html` key in the block representation for cases where there is a special HTML for a block instance, like when it has an error. And it was pointless to try to code a structure parsing the special ListBlock & StructBlock children when it might change in the future. So I made react-streamfield generic, with a simple structure for all block representations.

To conclude, we have two possible solutions for this:
- Write a database migration to migrate all StreamFields to a consistent structure. This could be very long on some sites, but the main reason I didn’t do it is that it needs its own discussion, and it’s risky. So it’s something we might end up doing, but in like 1 or 2 years
- Change the strucutre to make StreamBlock and ListBlock iterate over children to reformat their values. I guess that’s possible, but to be confirmed. I specifically remember trying to do it, and ending up with this instead. I will try again in the next days.

Reply:
> So as you can see, the poor FieldBlocks are not always considered as blocks in the database representation. The way they are treated depends on their parent. When a FieldBlock is in a StreamBlock, it’s considered as a normal block. When it’s in a ListBlock, it’s considered as a value. When it’s a StructBlock, it’s considered as a key/value pair.

No, that's not true - I think you've misunderstood the significance of the `{'id', 'type', 'value'}` container. This is the JSON equivalent of `StreamChild`, and is **an internal detail of StreamBlock** - no other block type will ever be required to handle it. It's not the thing that turns a block value into a "normal block". Or, to put it visually...

<img width="500" alt="streamfield-mental-model" src="https://user-images.githubusercontent.com/85097/54720867-72fe8980-4b58-11e9-95d5-08074ba8e4d8.png">

There is no inconsistency here: each block type decides on one value type to work with (which can be any Python data type), and consistently receives values of that exact type - not a wrapped version of it.

* for CharBlock, the value type is a string
* for PageChooserBlock, the value type is a `Page` object (which becomes an integer in the JSON representation)
* for ListBlock, the value type is a list of child values
* for StructBlock, the value type is a dict (or, more accurately, a `StructValue` which provides some extra semantics over `dict`) of child values
* for StreamBlock, the value type is a sequence of StreamChild-wrapped child values (which becomes a sequence of dict-wrapped child values in the JSON representation)

With a lot of persuasion, I could possibly come round to the idea of the `{'id', 'type', 'value'}` wrapper being understood and handled by each block as part of the base `Block` implementation, rather than being internal to StreamBlock. (Right now I'm not convinced that it's a worthwhile change: that wrapper dict consists of the metadata required for StreamBlock to do its job, and I can't see the same set of fields being relevant to other container types.) However, if we're doing that, we absolutely have to be explicit about the signature of each `Block` method, and whether it accepts/returns the wrapped or unwrapped version. Treating the two as interchangeable via type sniffing and undocumented `parent_block` logic is introducing vagueness to the block API, that I went to a lot of effort to avoid.

Reply:
Oh. You have to admit that this completely counter-intuitive.

Also, I stand to what I said, this makes the blocks representation inconsistent. With the current approach, `StreamBlock` has a very different representation from `ListBlock`, but they are the same thing. The only difference is that `ListBlock` can assume the type of its children. But that does not explain why the structure is so different. Also, this makes only `StreamBlock` store IDs, which is not great to identify blocks and version history.

As you said, there is no need to change the database representation for now. And it’s a good thing, because we can easily switch back to the old streamfield without having to write a forward **and** backwards migration.

There is one good reason to do it in the future though: simplification. We do not need `StreamChild`, `StructValue` or `ListBlock` with the approach I took, and we will be able to finally make a system where we expose blocks directly in templates. This would be a major improvement, as it would allow us to write block methods and handle blocks in a very similar fashion to models.

Reply:
A quick question before I respond in full: do you consider the "block representation" of a CharBlock to be

    "hello world"

or

    {"id": "98a82", "type": "heading", "value": "hello world"}

? The examples in your previous comment are ambiguous on this point, and I'd like to make sure we're using the same terminology.

Reply:
I understand that you consider that it is `"hello world"`, and I also understand that it is in a way what is currently implemented, but I consider that it would be clearer and easier to parse if we considered `{"id": "98a82", "type": "heading", "value": "hello world"}` to be the block representation.

This feels like the kind of things that made this possible in HTML5:
```
<ul>
    <li>First
    <li>Second
    <li>Third
</ul>
```
Obviously, it is “logic” that we do not have to close `li` tags because it’s an optimization that doesn’t introduce ambiguity. But on the other hand this makes it harder to understand HTML nowadays, and it makes it much harder to write a parser, because `li` is not a standalone tag, we are supposed to “guess” that the `li` gets closed by the next `li`…

It’s the same kind of feeling that I have here. OK, it’s more optimized, it may not be needed in all scenarii, but it adds complexity to parsing and understanding. Note as well that it makes it hard to cast streamfields to `JSON` or `JSONB` in PostgreSQL and search for a value. It’s probably the main reason why my streamfield collection system in https://github.com/wagtail/wagtail/pull/4702 was messy.

Reply:
Fixed by https://github.com/wagtail/wagtail/pull/4942/commits/bf862937ab62d08bf1f9f4037ce4567b4558dc85.